gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/domain_service/api/RollbackApiDomainServiceImpl.java

```diff
@@ -0,0 +1,240 @@
+import io.gravitee.apim.core.plan.domain_service.UpdatePlanDomainService;
+import io.gravitee.apim.core.plan.model.Plan;
+import io.gravitee.apim.core.plan.query_service.PlanQueryService;
+import io.gravitee.apim.infra.adapter.ApiAdapter;
+import io.gravitee.apim.infra.adapter.GraviteeJacksonMapper;
+import io.gravitee.apim.infra.adapter.PlanAdapter;
+import io.gravitee.common.utils.TimeProvider;
+import io.gravitee.definition.model.DefinitionVersion;
+import io.gravitee.definition.model.v4.plan.PlanStatus;
+import io.gravitee.rest.api.model.EventType;
+import io.gravitee.rest.api.service.common.ExecutionContext;
+import io.gravitee.rest.api.service.v4.ApiService;
+import java.time.ZonedDateTime;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@AllArgsConstructor
+@Service
+public class RollbackApiDomainServiceImpl implements RollbackApiDomainService {
+
+    private final AuditDomainService auditService;
+    private final EventCrudService eventCrudService;
+    private final PlanQueryService planQueryService;
+    private final CreatePlanDomainService createPlanDomainService;
+    private final UpdatePlanDomainService updatePlanDomainService;
+    private final ClosePlanDomainService closePlanDomainService;
+    private final ApiCrudService apiCrudService;
+
+    private final ApiService delegateApiService;
+
+    public void rollback(String eventId, AuditInfo auditInfo) {
+        var executionContext = new ExecutionContext(auditInfo.organizationId(), auditInfo.environmentId());
+
+        Event event = eventCrudService.get(auditInfo.organizationId(), auditInfo.environmentId(), eventId);
+
+        if (!event.getType().equals(EventType.PUBLISH_API) || event.getPayload() == null) {
+            throw new IllegalStateException("Cannot rollback an event that is not a publish event!");
+        }
+
+        try {
+            // Read API repository model from event
+            var apiRepositoryModel = GraviteeJacksonMapper
+                .getInstance()
+                .readValue(event.getPayload(), io.gravitee.repository.management.model.Api.class);
+
+            if (apiRepositoryModel.getDefinitionVersion() == null) {
+                throw new IllegalStateException("Cannot determine API definition version from event" + eventId);
+            }
+
+            if (apiRepositoryModel.getDefinitionVersion().equals(DefinitionVersion.FEDERATED)) {
+                throw new IllegalStateException("Cannot rollback a federated API");
+            }
+
+            if (!apiRepositoryModel.getDefinitionVersion().equals(DefinitionVersion.V4)) {
+                // TODO: Support rollback for V2 APIs
+                throw new IllegalStateException("Cannot rollback an API that is not a V4 API");
```

Comment:
two subject here (no ? ): 
> to compare the current version of the API to the version of the API in the event.
It's linked to another [PR](https://github.com/gravitee-io/gravitee-api-management/pull/6797) that isn't finished yet. I've added it to the sprint and added a comment to this one. I plan to continue on this in the next few days (if possible).

Reply:
> The case would be a v2 API migrated to V4, will we accept to rollback a V4 to a V2 ? (it's a question for the future)

No, I don't think this is an issue for rollback. because changes to a API v2 in V4 are not present only in the gateway. they are not tracked by the rollback. 
It's not a complete history of the API definition, including portal config, doc, ext...
it would be easier to speak by 📞 

but if we want to offer a way back after a V2 to V4 api migration, I'd go for the duplicate API during the V4 migration. or something like that.